Paper: itk-elastix: Medical image registration in Python

ntatsisk opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@deniederhut<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@TahiriNadia<!--end-reviewers-list-->

ntatsisk committed:
First draft

thewtex committed:
Updates and additions following review

ntatsisk committed:
Update based on feedback from the co-authors

ntatsisk committed:
Change the title of the paper

ntatsisk commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper ntatsisk-itk-elastix-paper scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

ntatsisk committed:
Consistent formating of elastix

ntatsisk commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper ntatsisk-itk-elastix-paper scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

ntatsisk committed:
Capitalize letter in title

ntatsisk commented:
Hi @deniederhut, I commited a couple of additional minor changes but I notice that the paper does not auto-build. Do you happen to know how to re-trigger it? Thanks!

deniederhut commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper ntatsisk-itk-elastix-paper scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

deniederhut commented:
@scoobies assign @deniederhut as editor

scoobies commented:
Assigned! @deniederhut is now the editor

deniederhut commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/S1361-8415(01)80026-8 is OK
- 10.1080/10255842.2012.670855 is OK
- 10.1109/TMI.2009.2035616 is OK
- 10.3389/fninf.2013.00050 is OK
- 10.1016/j.media.2010.10.003 is OK
- 10.1016/j.media.2015.12.004 is OK
- 10.3389/fninf.2014.00013 is OK
- 10.1109/CVPRW.2016.78 is OK
- 10.3389/fninf.2013.00045 is OK
- 10.48550/arXiv.2211.02701 is OK
- 10.48550/arXiv.2203.12362 is OK
- 10.5281/zenodo.5866555 is OK
- 10.1017/CBO9780511808432 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.5281/zenodo.3509134 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.1038/s41592-019-0686-2 is OK

MISSING DOIs

- None

INVALID DOIs

- /10.10162Fj.mri.2012.05.001 is INVALID
```

deniederhut commented:
@scoobies add @TahiriNadia to reviewers

scoobies commented:
@TahiriNadia added to the reviewers list!

deniederhut commented:
@scoobies code of conduct

scoobies commented:
Our CoC: https://www.scipy2023.scipy.org/code-of-conduct

scoobies commented:
Reports of abusive or harassing behavior may be reported to scipy@enthought.com

deniederhut commented:
Hey @TahiriNadia! Friendly reminder that your initial review is due June 23

TahiriNadia commented:
Hi @ntatsisk and your team, 

I am Nadia, one of your reviewers. The work is fascinating and impressive. I will provide some additional comments and questions for your article in a moment.

Cheers,
Nadia

TahiriNadia commented on papers/konstantinos_ntatsis/itkelastix.rst:
```diff
@@ -0,0 +1,275 @@
+
+The original and still-supported method to utilize ``elastix`` and ``transformix`` are command line executables. For the end user, this approach has the advantage that it does not require any external dependencies to be installed, which eases deployment. However, one limitation of this executable-based approach is its reliance on file input/output (I/O) operations. To address this limitation and  ...
+
+To further accommodate the needs of the users in the continuously developing scientific computing ecosystem, wrappings of the C++ code to other languages was developed in the form SimpleElastix :cite:`marstal2016simpleelastix`, which still exists as part of the SimpleITK :cite:`lowekamp2013design` package. More recently, we have embarked on developing a Python-specific wrapper called ``itk-elasti ...
```
Could you briefly describe the main features and capabilities of the ``itk-elastix`` Python package for recording medical images, and say how it compares with existing recording tools?

ntatsisk replied:
Hi @TahiriNadia, could you clarify the word "recording" in this comment? Do you maybe happen to mean "registration" instead? Thanks!

thewtex replied:
Hi @TahiriNadia @ntatsisk I added content for this in https://github.com/scipy-conference/scipy_proceedings/pull/820/commits/964a97445cd8c7421b1e453b2b1fe6ee41f00b15

TahiriNadia replied:
> Hi @TahiriNadia, could you clarify the word "recording" in this comment? Do you maybe happen to mean "registration" instead? Thanks!

Exactly.

TahiriNadia replied:
> Hi @TahiriNadia @ntatsisk I added content for this in [964a974](https://github.com/scipy-conference/scipy_proceedings/commit/964a97445cd8c7421b1e453b2b1fe6ee41f00b15)

thanks @thewtex.

TahiriNadia commented on papers/konstantinos_ntatsis/itkelastix.rst:
```diff
@@ -0,0 +1,275 @@
+--------------------------------
+The backend C++ ``elastix`` code is wrapped in Python with the Simplified Wrapper and Interface Generator (SWIG :cite:`swig`). The Python wrapping of ``elastix``, ``itk-elastix``, brings the power of ``elastix`` to the Python ecosystem, providing effortless integration with other scientific processing libraries and visualization software. The ``itk-elastix`` Python packages builds on the ``itk``  ...
+
+The process of updating and distributing the ``itk-elastix`` Python package is as follows: Once a significant number of changes have been made to the C++ ``elastix`` repository, a pull request is initiated in the ``itk-elastix`` repository to update its version. This triggers the ``itk-elastix`` Continuous Integration (CI) system, which performs builds of Python packages across various Python ver ...
```
It is mentioned that rigorous tests are carried out on ``elastix`` backend functionalities, with hundreds of tests performed on every pull request or commit. Is it possible to expand on this a little, mentioning for example some relevant categories of tests and their objectives?

TahiriNadia replied:
For consistency in nomenclature, would it be possible to replace elastix with ``elastix`` in the last sentence?

ntatsisk replied:
Yes certainly! Consistent use of `elastix` with https://github.com/scipy-conference/scipy_proceedings/pull/820/commits/ac51a15cc9d0690ee7686ef7bf29ab100d9e972c.

N-Dekker replied:
Thanks for your suggestion @TahiriNadia I think we could add something like the following:

> The test framework of elastix consists of various categories of tests, including low-level unit tests of the elastix library interface, minimal image registration tests on very small synthetic images, and larger regression tests of image registrations on realistic medical data. The tests are implemented using the CMake test driver CTest, the Python unittest module, and GoogleTest.

ntatsisk replied:
Thank you Niels, added via https://github.com/scipy-conference/scipy_proceedings/pull/820/commits/b02e581907a150a70717596643ddd74dfa8f1a52.

TahiriNadia replied:
Thank for your clarification.

TahiriNadia commented on papers/konstantinos_ntatsis/itkelastix.rst:
```diff
@@ -0,0 +1,275 @@
+
+.. class:: abstract
+
+Image registration plays a vital role in understanding changes that occur in 2D and 3D scientific imaging datasets. Registration involves finding a spatial transformation that aligns one image to another by optimizing relevant image similarity metrics. In this paper, we introduce ``itk-elastix``, a user-friendly Python wrapping of the mature ``elastix`` registration toolbox. The open-source tool  ...
```
Can ``itk-elastix`` effectively handle various imaging datasets, including rigid, affine, and B-spline deformable registration, and what are the advantages and limitations of each registration type? You can add a few lines on this in the section Image registration.

TahiriNadia replied:
It would be useful to know more about similarity metrics?

ntatsisk replied:
Thanks for the comment. Added details on parametric models (https://github.com/scipy-conference/scipy_proceedings/pull/820/commits/99d5183b53815e5b438dba3adb1e168ba7e3f25c) and on similarity metrics (https://github.com/scipy-conference/scipy_proceedings/pull/820/commits/d32c38afa9d01bfed2c5be9c13734d923b9e83b3).

TahiriNadia replied:
@ntatsisk: Thank you for these improvements.

TahiriNadia commented on papers/konstantinos_ntatsis/itkelastix.rst:
```diff
@@ -0,0 +1,275 @@
+Introduction
+------------
+
+Image Registration
```
Can you elaborate on why this approach is specifically intriguing or advantageous in terms of enhancing the quality of MRI brain image data? Can it not be generalized to other types of images?

ntatsisk replied:
Hi @TahiriNadia, `itk-elastix` is by no means bound to work only with brain MRI images. It works well also with CT, PET, Ultrasound, X-ray etc. and also for different anatomical regions (head, liver, heart, knee, prostate etc). Our model zoo contains several of those examples https://elastix.lumc.nl/modelzoo/.

Could you please pin-point where it is implied otherwise? I will try to correct the text. Unfortunately, at the moment I can't see where you exactly placed the comment. Thanks!

TahiriNadia replied:
The introduction section is the perfect location to effectively clarify several key points.

TahiriNadia commented on papers/konstantinos_ntatsis/itkelastix.rst:
```diff
@@ -0,0 +1,275 @@
+------------------------------------
+ITK Transforms
+++++++++++++++
+In addition to the fact that ``elastix`` is based on ITK, there is an ongoing effort to increase the compatibility between the two libraries even further. One particular example is the Transform classes. In the following example, we show that ITK transforms can be used directly by ``transformix``:
```
Could you please include the following reference to invite readers to learn more about the foundation for assessing image registration? 

Avants, B. B., Tustison, N. J., Stauffer, M., Song, G., Wu, B., & Gee, J. C. (2014). The Insight ToolKit image registration framework. Frontiers in neuroinformatics, 8, 44.
doi: [10.3389/fninf.2014.00044](https://doi.org/10.3389%2Ffninf.2014.00044)

ntatsisk replied:
Good idea. Added via https://github.com/scipy-conference/scipy_proceedings/pull/820/commits/6787e7fe7b2c99fa0c4263bd38b233ddbe87348e.

TahiriNadia replied:
Thanks @ntatsisk

TahiriNadia commented on papers/konstantinos_ntatsis/itkelastix.rst:
```diff
@@ -0,0 +1,275 @@
+
+Project MONAI
++++++++++++++
+More and more people work on the application of deep learning to medical imaging research. To that end, we developed `itk_torch_bridge`__ as module of the MONAI codebase that allows conversion 1) of an ITK Image to a MONAI MetaTensor and the reverse, while making sure that all relevant metadata remain intact, and 2) an ITK Transform to a MONAI Transform and back. The latter is necessary since the ...
```
Sometimes ITK Image and other times ITK *Image*

ntatsisk replied:
Thanks, addressed with https://github.com/scipy-conference/scipy_proceedings/pull/820/commits/68c617a037b6c6d028ef93c00513008301828827.

TahiriNadia replied:
Thanks @ntatsisk.

TahiriNadia commented on papers/konstantinos_ntatsis/itkelastix.rst:
```diff
@@ -0,0 +1,275 @@
+
+   The user interface of the ``elastix-napari`` plugin. For a larger version of the image: https://github.com/SuperElastix/elastix-napari#elastix-napari. :label:`elastix-napari`
+
+Continuous efforts have been made to make ``itk-elastix`` accessible to users of various tools. One notable community-driven initiative is SlicerElastix, which seamlessly integrates ``elastix`` (as an executable) into 3D Slicer :cite:`fedorov20123d` medical image visualization software. In addition to this, recent endeavors focused on developing the ``elastix-napari`` plugin for the Napari :cite: ...
```
It would be interesting to add the following reference for the Napari user interface.

Windhager, J., Bodenmiller, B., & Eling, N. (2021). An end-to-end workflow for multiplexed image processing and analysis. BioRxiv, 2021-11.
doi: https://doi.org/10.1101/2021.11.12.468357

ntatsisk replied:
Thanks @TahiriNadia. I took a deeper look in the publication you shared here. My understanding is that the authors propose a workflow for the analysis of multiplexed images (cell images). One part of that workflow regards reading and visualizing the images using a napari plugin named `napari-imc` (https://github.com/BodenmillerGroup/napari-imc) developed by the authors of the paper that you shared. In a similar manner, we developed the `elastix-napari` plugin (https://github.com/SuperElastix/elastix-napari) described in this under-review paper. 

So, if my understanding is correct, those authors are not affilated in any way with the Napari development team and thus we can't use that citation for the Napari tool. However, I did find in the Napari official repository that the proper way to cite the software is this: https://github.com/napari/napari#citing-napari. So, I corrected the reference that we were using by commit https://github.com/scipy-conference/scipy_proceedings/pull/820/commits/abde598e8ef41ed0f7da5fc37c524f91af9e151e. Please, let me know if I have misunderstood your comment.

TahiriNadia replied:
That's precisely what I intended to convey. You have provided a comprehensive answer to my question. Thank you, @ntatsisk,  for your attentive response.

TahiriNadia commented on papers/konstantinos_ntatsis/itkelastix.rst:
```diff
@@ -0,0 +1,275 @@
+
+Concluding remarks
+------------------
+We presented ``itk-elastix``, an easy-to-install and easy-to-use Python package that lowers the entry barrier for multi-dimensional image registration. Its key features are 1) a robust and well-established backend codebase that provides stability and reliability, 2) an extensive collection of tutorials, a parameter file model zoo, and up-to-date documentation as comprehensive resources for user a ...
```
Is it useful to provide guidance on the next steps to be taken and explain the reasons why?

thewtex replied:
@TahiriNadia yes, good idea.

@ntatsisk , we could include, e.g.

> Next steps will improve universal access to `itk-elastix` and make end-to-end deep learning segmentation and registration pipelines applicable to diverse medical datasets. These steps include a port to WebAssembly and utility Python deep learning resampling functions.

ntatsisk replied:
Thanks Matt, I rephrased a bit your suggestion in https://github.com/scipy-conference/scipy_proceedings/pull/820/commits/262bbf75f368dadae99c540b91906acd3762e713.

TahiriNadia replied:
@ntatsisk and @thewtex: Thank you both for your work.

deniederhut commented:
@scoobies mark pending comment

deniederhut commented:
Hi @ntatsisk ! How are the revisions coming along?

ntatsisk committed:
Fix typo in code snippet

ntatsisk committed:
Use ITK Image and Transform consistently

Adressing https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239327404

ntatsisk committed:
Write ``elastix`` consistently throughout the text

To address comment https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239313347

ntatsisk commented:
Hi @TahiriNadia, many thanks for the kind words and the feedback! We will get to work to address your comments. :)

Hi @deniederhut, thanks for asking! Actually, I had taken a few days off but I will start making the revisions right away.

ntatsisk committed:
Add parametric registration description in intro

ntatsisk committed:
Add information on similarity metrics

For comment:
https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239338519

ntatsisk committed:
Add reference to ITK registration

For comment:
https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239323968

ntatsisk committed:
Fix invalid 3D Slicer DOI

ntatsisk committed:
Correct the Napari citation

deniederhut commented:
Okay great! Feel free to @ Nadia when this is ready for a re-review

ntatsisk committed:
Add Next Steps in conclusion

Related to comment:
https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239348648

ntatsisk committed:
Fix typos

ntatsisk committed:
Explain further the testing framework

As suggested by Niels Dekker. Adresses:
https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239312503

ntatsisk committed:
Improve Figure 1 position in text

ntatsisk committed:
Fix alignment in code snippet

ntatsisk commented:
Hi @TahiriNadia, thank you again for your review.  We tried to address most of your comments. The ones remaining that might need some clarification first are: https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239310037 and https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239317578. Let us know what you think.

thewtex committed:
Summarize main features

TahiriNadia commented:
> Hi @TahiriNadia, thank you again for your review. We tried to address most of your comments. The ones remaining that might need some clarification first are: [#820 (comment)](https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239310037) and [#820 (comment)](https://github.com/scipy-conference/scipy_proceedings/pull/820#discussion_r1239317578). Let us know what you think.

Thank you very much, @ntatsisk and @thewtex, for your hard work and for graciously considering my suggestions.

ntatsisk commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper ntatsisk-itk-elastix-paper scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

deniederhut commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/JPROC.2003.817864 is OK
- 10.1016/S1361-8415(01)80026-8 is OK
- 10.1080/10255842.2012.670855 is OK
- 10.1109/TMI.2009.2035616 is OK
- 10.3389/fninf.2013.00050 is OK
- 10.1016/j.media.2010.10.003 is OK
- 10.1016/j.media.2015.12.004 is OK
- 10.3389/fninf.2014.00013 is OK
- 10.1109/CVPRW.2016.78 is OK
- 10.3389/fninf.2013.00045 is OK
- 10.48550/arXiv.2211.02701 is OK
- 10.48550/arXiv.2203.12362 is OK
- 10.5281/zenodo.3555620 is OK
- 10.3389/fninf.2014.00044 is OK
- 10.1016/j.mri.2012.05.001 is OK
- 10.1017/CBO9780511808432 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1038/s41586-020-2649-2 is OK
- 10.5281/zenodo.3509134 is OK
- 10.25080/Majora-92bf1922-00a is OK
- 10.1038/s41592-019-0686-2 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

deniederhut commented:
I skimmed this paper very quickly so please let me know if I'm wrong about this -- but I see no references to freesurfer, FSL, or SPM, all of which have implementations of cranial image registration. We care a lot about referencing prior art at SciPy so if this is not in the paper it should be added before publication.

ntatsisk commented:
Thank you @deniederhut for taking a look at our paper and providing feedback. 

You are correct that there are no references of freesurfer, FSL or SPM in the paper. However, I feel that your comment implies that ``elastix`` (and hence ``itk-elastix`` python wrapping) is a specialized cranial registration tool, which is not the case. In the contrary, it is a general medical image registration toolbox that works for different anatomical regions - that is explained inside the paper, but you can also glance through the collection of community applications here: https://elastix.lumc.nl/modelzoo/ to see that they are not restricted to brain registration. We simply happened to choose brain images as registration example but we could just as easily had chosen something else e.g. lungs.

We also do care a lot for referencing related art but only as long as it serves the reader and does justice to the people that we based our work on. So, publications directly related to our work are included, but we cannot possibly include all registration methods or tools out there since the research field is vast. The listing of such publications is better suited for review or survey papers - we have included a sufficient number of those in the introduction so that the interested reader can delve further into the subject.

I hope this comment addresses your concerns. Please let me know if that is not the case, or if you have further feedback.

deniederhut commented:
> We simply happened to choose brain images as registration example

As a former brain person I approve! Thanks for the clarification.

deniederhut commented:
@scoobies mark ready

deniederhut reviewed (approved):
Thanks @TahiriNadia for helping us get this paper ready for publication! And thanks @ntatsisk for joining us at SciPy this year!

![giphy-1822950437](https://github.com/scipy-conference/scipy_proceedings/assets/6868396/855f010c-c978-4cac-bf43-3e5faf45fd21)
